File: linear-search/linear-search.py
def linear_search(arr, target):
    for i, val in enumerate(arr):
        if val == target:
            return i
    return -1

def linear_search_recursive(arr, target, index=0):
    if index >= len(arr):
        return -1
    if arr[index] == target:
        return index
    return linear_search_recursive(arr, target, index + 1)

print(linear_search([1, 2, 3, 4, 5], 3))
print(linear_search_recursive([1, 2, 3, 4, 5], 3))

File: linear-search/linear-search.rs
fn linear_search(arr: &[i32], target: i32) -> Option<usize> {
    for (i, &val) in arr.iter().enumerate() {
        if val == target { return Some(i); }
    }
    None
}

fn linear_search_recursive(arr: &[i32], target: i32, index: usize) -> Option<usize> {
    if index >= arr.len() { return None; }
    if arr[index] == target { return Some(index); }
    linear_search_recursive(arr, target, index + 1)
}

fn main() {
    let arr = [1, 2, 3, 4, 5];
    println!("{:?}", linear_search(&arr, 3));
    println!("{:?}", linear_search_recursive(&arr, 3, 0));
}